File: server/internal/services/character_service.go
package services

import (
	"errors"
	"warhammer-api/internal/models"
	"warhammer-api/internal/repositories"
)

type CharacterService struct {
	repo repositories.CharacterRepository
}

func NewCharacterService(repo repositories.CharacterRepository) *CharacterService {
	return &CharacterService{repo}
}

func (s *CharacterService) GetAll() ([]models.Character, error) {
	return s.repo.GetAll()
}

func (s *CharacterService) GetByID(id int) (*models.Character, error) {
	return s.repo.GetById(id)
}

func (s *CharacterService) Create(data *models.Character) (*models.Character, error) {
	if data.Name == "" {
		return nil, errors.New("name cannot be empty")
	}
	return s.repo.Create(data)
}

func (s *CharacterService) Update(data *models.Character) (*models.Character, error) {
	return s.repo.Update(data)
}

func (s *CharacterService) Delete(id int) error {
	return s.repo.Delete(id)
}

File: server/services/characterService.js
const characterRepository = require("../repositories/characterRepository");
const { createLegacyNotFoundError } = require("../lib/legacyApi");

class CharacterService {
  async getAll(options = {}) {
    return await characterRepository.findAll(options);
  }

  async getById(id) {
    const character = await characterRepository.findById(id);
    if (!character) {
      throw createLegacyNotFoundError("Character");
    }
    return character;
  }

  async create(characterData) {
    return await characterRepository.create(characterData);
  }

  async update(id, characterData) {
    const updated = await characterRepository.update(id, characterData);
    if (!updated) {
      throw createLegacyNotFoundError("Character");
    }
    return updated;
  }

  async delete(id) {
    const deleted = await characterRepository.delete(id);
    if (!deleted) {
      throw createLegacyNotFoundError("Character");
    }
    return deleted;
  }
}

module.exports = new CharacterService();
